src/packages/recompose/withPropsOnChange.js

```diff
@@ -1,34 +1,37 @@
-import { Component } from 'react'
 import pick from './utils/pick'
 import shallowEqual from './shallowEqual'
 import createHelper from './createHelper'
-import createEagerFactory from './createEagerFactory'
+import createHocFromMiddleware from './utils/createHocFromMiddleware'
 
-const withPropsOnChange = (shouldMapOrKeys, propsMapper) => BaseComponent => {
-  const factory = createEagerFactory(BaseComponent)
+const withPropsOnChange = (shouldMapOrKeys, propsMapper) => {
   const shouldMap = typeof shouldMapOrKeys === 'function'
     ? shouldMapOrKeys
     : (props, nextProps) => !shallowEqual(
         pick(props, shouldMapOrKeys),
         pick(nextProps, shouldMapOrKeys),
       )
 
-  return class extends Component {
-    computedProps = propsMapper(this.props);
+  return createHocFromMiddleware(({ getProps }) => next => {
+    let didUpdate = false
+    let computedProps = null
 
-    componentWillReceiveProps(nextProps) {
-      if (shouldMap(this.props, nextProps)) {
-        this.computedProps = propsMapper(nextProps)
-      }
+    const update = (props, cb) => {
```

Comment:
No need to create a `update` function here.
We can move `next.update(...)` to line 31.